File: python-refactor/Code/pre_03_station/KR_station_01_read_raw.m
clear all; close all; clc

tic

path_insiu = '//10.72.26.46/irisnas6/Data/In_situ/AirQuality_SouthKorea/';
path = '//10.72.26.56/irisnas5/Data/Station/Station_Korea/';

% path_data = '/share/irisnas2/Data/Aerosol/';
% path = '/share/irisnas5/Data/Station/Station_Korea/';

% addpath(genpath('/share/irisnas5/Data/matlab_func/'))

%% 2005-2013 (xlsx, PM2.5없음)
for yr=2005:2013 
data = [];
for i=1:4
    [num,txt,raw] = xlsread([path_insiu,num2str(yr),'/',num2str(yr),'년',num2str(i,'%02i'),'분기.xlsx']);
    data_temp = txt(2:end,3:9);
    data = [data;data_temp];
end

dvec = datevec(data(:,2),'yyyymmddHH');
data_datenum = datenum(data(:,2),'yyyymmddHH');
doy_000 = datenum([yr,0,0,0,0,0]);
data_doy = floor(data_datenum-doy_000);

data = str2double(data);
data(data<0)=NaN;

ndata = [data_doy,dvec(:,1:4),data(:,3:end),NaN(size(data_doy)),data(:,1)];
% doy, yr, mon, day, time, SO2, CO, O3, NO2, PM10, (PM25), scode

save([path,'stn_code_data/stn_code_data_',num2str(yr)],'ndata','-v7.3')
end


%% 2014-2016 (csv, PM2.5 컬럼 있음)
for yr=2014:2016 
data = [];
for i=1:4
    [num,txt,raw] = xlsread([path_insiu,num2str(yr),'/',num2str(yr),'년',num2str(i),'분기.csv']);
    data_temp = raw(2:end,[2,4:10]);
    data = [data;data_temp];
end

data = cell2mat(data);
data(data<0)=NaN;

dstr = num2str(data(:,2));
dvec = datevec(dstr,'yyyymmddHH');
data_datenum = datenum(dstr,'yyyymmddHH');
doy_000 = datenum([yr,0,0,0,0,0]);
data_doy = floor(data_datenum-doy_000);

ndata = [data_doy,dvec(:,1:4),data(:,3:end),data(:,1)];
% doy, yr, mon, day, time, SO2, CO, O3, NO2, PM10, (PM25), scode

save([path,'stn_code_data/stn_code_data_',num2str(yr)],'ndata','-v7.3')
end

%% 2017-2018 2분기 (xlsx, PM2.5 있음)
for yr=2018 %2017:2018
data = [];
for i=1:4
    [num,txt,raw] = xlsread([path_insiu,num2str(yr),'/',num2str(yr),'년 ',num2str(i),'분기.xlsx']);
    data_temp = raw(2:end,[2,4:10]);
    data = [data;data_temp];
end

data_1 =  str2double(data(:,1));
data_2 = cell2mat(data(:,2:end));
data = [data_1, data_2];
data(data<0)=NaN;

dstr = num2str(data(:,2));
dvec = datevec(dstr,'yyyymmddHH');
data_datenum = datenum(dstr,'yyyymmddHH');
doy_000 = datenum([yr,0,0,0,0,0]);
data_doy = floor(data_datenum-doy_000);

ndata = [data_doy,dvec(:,1:4),data(:,3:end),data(:,1)];
% doy, yr, mon, day, time, SO2, CO, O3, NO2, PM10, (PM25), scode

save([path,'stn_code_data/stn_code_data_',num2str(yr)],'ndata','-v7.3')
end

%% 2018 3분기 - 4분기 (xlsx, PM2.5있음.. 근데 중간에 망 정보가 들어가면서 컬럼 위치가 변경됨)
i=4;
[num,txt,raw] = xlsread([path_insiu,num2str(yr),'/',num2str(yr),'년 ',num2str(i),'분기.xlsx']);
data_temp = raw(2:end,[3,5:11]);
data = [data;data_temp];

%% 201810_201904 
data = [];
[num,txt,raw] = xlsread([path_insiu,'201810_201904/201810_201904_pt1.csv']);
data_temp = raw(2:end,[2,6:7,10,9,8,11,12]);
data = [data;data_temp];

[num,txt,raw] = xlsread([path_insiu,'201810_201904/201810_201904_pt2.csv']);
data_temp = raw(2:end,[2,6:7,10,9,8,11,12]);
data = [data;data_temp];

data = cell2mat(data);
data(data<0)=NaN;

dstr = num2str(data(:,2));
dvec = datevec(dstr,'yyyymmddHH');
yr=2019;
yr_idx = dvec(:,1)==yr;
dvec=dvec(yr_idx,:);
dstr=dstr(yr_idx,:);
data=data(yr_idx,:);
data_datenum = datenum(dstr,'yyyymmddHH');

doy_000 = datenum([yr,0,0,0,0,0]);
data_doy = floor(data_datenum-doy_000);

ndata = [data_doy,dvec(:,1:4),data(:,3:end),data(:,1)];
% doy, yr, mon, day, time, SO2, CO, O3, NO2, PM10, (PM25), scode

save([path,'stn_code_data/stn_code_data_',num2str(yr)],'ndata','-v7.3')


%% 12월 31일 24시 -> 1월 1일 00시 
path = '/share/irisnas5/Data/Station/Station_Korea/';

yr=2005;
load([path,'stn_code_data/stn_code_data_',num2str(yr)]);
data_mv = ndata(ndata(:,2)==(yr+1),:);
doy_unq = unique(data_mv(:,1))

% if length(doy_unq)==1
    data_mv(:,1)=1;
% else
%     stop
% end

ndata(ndata(:,2)==(yr+1),:)=[];
save([path,'stn_code_data/stn_code_data_',num2str(yr)],'ndata','-v7.3')

for yr=2016:2017 %2006:2017
load([path,'stn_code_data/stn_code_data_',num2str(yr)]);
ndata = [data_mv; ndata];
data_mv = ndata(ndata(:,2)==(yr+1),:);
doy_unq = unique(data_mv(:,1))

if length(doy_unq)==1
    data_mv(:,1)=1;
    ndata(ndata(:,2)==(yr+1),:)=[];
    save([path,'stn_code_data/stn_code_data_',num2str(yr)],'ndata','-v7.3')
else
    stop
end

end

save([path,'stn_code_data/stn_code_data_',num2str(yr),'_001_00'],'data_mv','-v7.3')

File: python-refactor/Code/pre_03_station/KR_station_01_read_raw.py
### Package Import
import sys
import os
base_dir = os.environ['GEMS_HOME']
project_path = base_dir
sys.path.insert(0, project_path)
from Code.utils import matlab

import copy
import numpy as np 
import pandas as pd
import glob

### Setting path
data_base_dir = os.path.join(base_dir, 'Data')
path_in_situ = os.path.join(data_base_dir, 'Raw', 'In_situ')
path_station = os.path.join(data_base_dir, 'Station') 
path_stn_kr = os.path.join(path_station, 'Station_KR')

cols=['측정소코드','측정일시','SO2','CO','O3','NO2','PM10','PM25']

## 
YEARS = range(2014, 2018+1)
for yr in YEARS:
    flist = glob.glob(os.path.join(path_in_situ,'AirQuality_SouthKorea',str(yr),'*'))
    flist.sort()
    data = None
    for fname in flist:
        fname_body, fname_ext = os.path.splitext(fname)
        if fname_ext=='.csv':
            try:
                tmp_df = pd.read_csv(fname)
            except:
                tmp_df = pd.read_csv(fname,encoding='euc-kr')
        else:
            tmp_df = pd.read_excel(fname)

        data_temp = tmp_df.loc[:,cols]
        if data is None:
            data = data_temp
        else:
            data = pd.concat([data, data_temp])
        del tmp_df, data_temp
    data = data.values
    data[data<0]=np.nan

    dstr = [str(int(x)) for x in data[:,1]]
    tmp_df = pd.DataFrame(dstr, columns=['time'])
    pat = '(?P<year>\d{4})(?P<month>\d{2})(?P<day>\d{2})(?P<hour>\d{2})'
    dvec = pd.to_datetime(tmp_df['time'].str.extract(pat, expand=True))
    dstr = dvec.dt.strftime('%Y%m%d').values
    data_datenum = [matlab.datenum(val) for val in dstr]
    doy_000 = matlab.datenum(f'{yr}00000')
    data_doy = np.array([np.floor(val-doy_000) for val in data_datenum])
    dvec = [(val.year, val.month, val.day, val.hour) for val in dvec]
    dvec = np.array(dvec)
    ndata = np.hstack([data_doy.reshape(-1,1),dvec,data[:,2:],data[:,0].reshape(-1,1)])
    # doy, yr, mon, day, time, SO2, CO, O3, NO2, PM10, PM25, scode
    matlab.savemat(os.path.join(path_stn_kr,'stn_code_data',f'stn_code_data_{yr}.mat'), 
                    {'ndata':ndata})

## Dec 31 24:00 KST -> Jan 1 00:00 KST
for yr in YEARS:
    fname = f'stn_code_data_{yr}.mat'
    ndata = matlab.loadmat(os.path.join(path_stn_kr,'stn_code_data', fname))['ndata']
    fname_mv = f'stn_code_data_{yr}_001_00.mat'
    if yr!=YEARS[0]:
        data_mv = matlab.loadmat(os.path.join(path_stn_kr,'stn_code_data', fname_mv))['data_mv']
        ndata = np.vstack([data_mv, ndata])
        del data_mv
    data_mv = ndata[ndata[:,1]==(yr+1),:]
    doy_unq = np.unique(data_mv[:,0])

    if len(doy_unq)==1:
        data_mv[:,0]=1
        idx= ndata[:,1]==(yr+1)
        ndata = ndata[~idx]
        matlab.savemat(os.path.join(path_stn_kr,'stn_code_data', fname), {'ndata':ndata})
        if yr!=YEARS[0]:
            os.remove(os.path.join(path_stn_kr,'stn_code_data', fname_mv))
    else:
        raise ValueError('len(doy_unq)!=1')

    fname_mv2 = f'stn_code_data_{yr+1}_001_00.mat'
    matlab.savemat(os.path.join(path_stn_kr,'stn_code_data', fname_mv2), {'data_mv':data_mv})
